crema/confidence.py

```diff
@@ -346,3 +440,158 @@ def _assign_confidence(self):
 
             self.confidence_estimates[level] = df.loc[targets, :]
             self.decoy_confidence_estimates[level] = df.loc[~targets, :]
+
+
+class MixmaxConfidence(Confidence):
+    """Assign confidence estimates using mix-max competition.
+
+    #TODO. Maybe just cite paper
+    Estimates qvalues using the mix-max competition method. For a set
+    of ..., the false discover rate (FDR) is estimated.
+
+    Parameters
+    ----------
+    psms : a PsmDataset object
+        A collection of PSMs
+    score_column : str, optional
+        The score by which to rank the PSMs for confidence estimation. If
+        :code:`None`, the score that yields the most PSMs at the specified
+        false discovery rate threshold (`eval_fdr`), will be used.
+    desc : bool, optional
+        True if higher scores better, False if lower scores are better.
+        If None, crema will try both and use the
+        choice that yields the most PSMs at the specified false discovery
+        rate threshold (`eval_fdr`). If `score_column` is :code:`None`,
+        this parameter is ignored.
+    eval_fdr : float, optional
+        The false discovery rate threshold used to evaluate the best
+        `score_column` and `desc` to choose. This should range from 0 to 1.
+    pep_fdr_type : {"psm-only","peptide-only",psm-peptide"}, optional
+        The method for crema to use when calculating peptide level confidence
+        estimates.
+
+    Attributes
+    ----------
+    data : pandas.DataFrame
+    dataset : crema.PsmDataset
+    levels : list of str
+    confidence_estimates : Dict
+        A dictionary containing the confidence estimates at each level, each
+        as a :py:class:`pandas.DataFrame`.
+    decoy_confidence_estimates : Dict
+        A dictionary containing the confidence estimates for the decoy hits at
+        each level, each as a :py:class:`pandas.DataFrame`
+    """
+
+    def __init__(
+        self,
+        psms,
+        score_column=None,
+        desc=None,
+        eval_fdr=0.01,
+        pep_fdr_type="psm-peptide",
+    ):
+        """Initialize a TdcConfidence object."""
+        LOGGER.info(
+            "Assigning confidence estimates using mix-max competition..."
+        )
+
+        # TODO maybe better way to do this
+        # can not infer desc value as the wrong value will
+        # result in a divide by zero error
```

Comment:
TBH I implemented mixmax without fully understanding how it works. I copied the Percolator code and did a system check to make sure it was correct. I need to read the paper again

However, essentially the desc parameter indicates whether larger or smaller scores as better. qvalues.py::calculate_mixmax_qval wants scores sorted from worst to best. Inside that fxn, there are some divisions involved. If the desc parameter does not match the score (ie XCorr with smaller is better) it will cause the denom of the division to be zero. ie a divide by 0 error.

Did that make sense @wfondrie ?

Reply:
Yep that makes sense!